src/game_loop.c

```diff
@@ -142,11 +146,14 @@ void process_dungeon_destroy(struct Thing* heartng)
             }
             else if (dungeon->heart_destroy_turn == 28)
             {
-                terminate_thing_spell_effect(soultng, SplK_Invisibility);
+                // Strange case, not sure why it's required.
+                sctrl = creature_control_get_from_thing(soultng);
+                clear_flag(sctrl->spell_flags, CSAfF_Invisibility);
+                sctrl->force_visible = 0;
             }
             else if (dungeon->heart_destroy_turn == 30)
-            {
-                dungeon->free_soul_idx = 0; 
+            { // 'soultng' is deleted shortly after so why even bother making it invisible with spell flags? Perhaps it could use rendering flag instead?
```

Comment:
This is when you have a backup heart. When the first heart is destroyed, the soul of the keeper will leave it. It will then teleport to the next heart and activate that. Making it invisible is a visual effect to indicate it teleporting.

Reply:
Would it still become invisible? If so, then yes.

Reply:
ok, good.

Reply:
Then you can remove your 'I do not understand' comments, or replace them with an explanation if you feel there is a need.